api/client/gro_client.py

```diff
@@ -370,22 +370,37 @@ def find_data_series(self, **kwargs):
             search_results.append(
                 self.search('regions', kwargs['partner_region'])[:cfg.MAX_RESULT_COMBINATION_DEPTH])
             keys.append('partner_region_id')
-        all_data_series = []
+        # Rank by frequency and source, while preserving search ranking in
+        # permutations of item, metric, region, and partner region.
+        freq_ranking = []
+        ranking_groups = set()
+        count = 0
         for comb in itertools.product(*search_results):
             entities = dict(list(zip(keys, [entity['id'] for entity in comb])))
-            data_series_list = self.get_data_series(**entities)
-            self._logger.debug("Found {} distinct data series for {}".format(
-                len(data_series_list), entities))
-            # temporal coverage affects ranking so add time range if specified.
-            for data_series in data_series_list:
-                if kwargs.get('start_date'):
-                    data_series['start_date'] = kwargs['start_date']
-                if kwargs.get('end_date'):
-                    data_series['end_date'] = kwargs['end_date']
-            all_data_series += data_series_list
-        self._logger.info("Found {} distinct data series total for {}".format(
-            len(all_data_series), kwargs))
-        for data_series in self.rank_series_by_source(all_data_series):
+            for data_series in self.get_data_series(**entities)[:cfg.MAX_SERIES_PER_COMB]:
+                count += 1
+                self._logger.debug("Data series: {}".format(data_series))
+                # time range affects ranking
+                data_series.pop('start_date', None)
+                data_series.pop('end_date', None)
+                data_series.pop('frequency_id', None)
+                data_series.pop('source_id', None)
+                data_series.pop('source_name', None)
+                data_series.pop('metadata', None)
+                series_hash = frozenset(data_series.items())
+                if series_hash not in ranking_groups:
+                    ranking_groups.add(series_hash)
+                    if kwargs.get('start_date'):
+                        data_series['start_date'] = kwargs['start_date']
+                    if kwargs.get('end_date'):
+                        data_series['end_date'] = kwargs['end_date']
+                    for tf in self.get_available_timefrequency(**data_series):
+                        ds = dict(data_series)
+                        ds['frequency_id'] = tf['frequency_id']
+                        freq_ranking.append(ds)
```

Comment:
I don't think we need this `freq_ranking` list at all.

If instead of `freq_ranking.append(ds)` you just do

```py
for data_series in self.rank_series_by_source([ds])
  yield data_series
```

you'd get the same results since in `rank_series_by_source` it's just going through the input list in order and only ranking sources for one input series at a time. Not re-ordering all the results based on source ranking.

That way, the generator can start yielding results before going through all of the combinations first. That would address my above complaint about `MAX_SERIES_PER_COMB` being 1000 taking a long time.

Reply:
yes  i was planning on doing that in the first place but then decided to  log the intermediate results before source ranking...

Reply:
There's actually a difference in theory though not in practice. We are ranking all series by time and frequency and then re-ranking  by source. So the source ranking could reshuffle the frequencies. E.g. timefreq gives

(freq 1, region1), (freq2, region2) ... 

and then source ranking gives:

(freq2, region2, source1),  (freq1, region1, source2),  

the current implementation of rank_series_by_source does not do that, but we might want that.

Reply:
changed. right now I think the simplication is worth it.